src/abi_win32.cpp

```diff
@@ -42,34 +42,34 @@ AbiState default_abi_state = 0;
 
 bool use_sret(AbiState *state, jl_value_t *ty)
 {
-    if (!jl_is_datatype(ty) || jl_is_abstracttype(ty) || jl_is_cpointer_type(ty) || jl_is_array_type(ty))
-        return false;
-    size_t size = jl_datatype_size(ty);
-    if (size <= 8)
+    // Assume jl_is_datatype(ty) && !jl_is_abstracttype(ty)
+    jl_datatype_t *dt = (jl_datatype_t*)ty;
+    // Use sret if the size of the argument is not one of 1, 2, 4, 8 bytes
+    // This covers the special case of Complex64
+    size_t size = dt->size;
+    if (size == 1 || size == 2 || size == 4 || size == 8)
         return false;
     return true;
 }
 
 void needPassByRef(AbiState *state, jl_value_t *ty, bool *byRef, bool *inReg)
 {
-    if (!jl_is_datatype(ty) || jl_is_abstracttype(ty) || jl_is_cpointer_type(ty) || jl_is_array_type(ty))
-        return;
-    size_t size = jl_datatype_size(ty);
-    if (size <= 8)
-        return;
-    *byRef = true;
+    // Assume jl_is_datatype(ty) && !jl_is_abstracttype(ty)
+    jl_datatype_t *dt = (jl_datatype_t*)ty;
+    // Use pass by reference for all structs
+    *byRef = dt->nfields > 0;
```

Comment:
> size == 16 and <: FloatingPoint

we may want to revisit that as well. in particular, that might accidentally include a non-IEEE type (such as an mpfr-like value). the other cases are handled explicitly (with comparisons to Float64, etc).

> The Int128 test should actually be fixed (or comment about) for aarch64 since it is actually a c type. Probably a different PR.

sounds good. my original ccall test code was making lots of shortcuts based upon how i knew the x86-family ABIs would behave